jl4-lsp/src/LSP/L4/Viz/CustomProtocol.hs

```diff
@@ -1,48 +1,78 @@
 {-# LANGUAGE DataKinds #-}
 
 {-| Custom extensions to the LSP and payload types for JL4
-    See also @ts-shared/jl4-client-rpc/custom-protocol.ts@
+  See also 
+  - @ts-shared/jl4-client-rpc/custom-protocol.ts@ for RPC-related TS types
+  - @ts-shared/viz-expr/eval-on-backend.ts@ for the params/response payload TS types
 -}
 module LSP.L4.Viz.CustomProtocol where
 
 import qualified Base.Text as T
-import Data.Aeson   (FromJSON, ToJSON)
+import Data.Aeson (FromJSON(..), ToJSON(..), object, (.:), (.=), withObject)
 import GHC.Generics (Generic)
 import Data.Proxy (Proxy(..))
 import GHC.TypeLits  (Symbol, KnownSymbol, symbolVal)
 import Language.LSP.Protocol.Types as LSP
-import LSP.L4.Viz.VizExpr
+import LSP.L4.Viz.VizExpr as V
+import GHC.Records (HasField)
 
 ------------------------------------------------------
 -- l4/evalApp Request Params and Response Payload
 ------------------------------------------------------
 
 -- | Payload / params for EvalAppRequest
 data EvalAppRequestParams = EvalAppRequestParams
-  { appExpr :: ID,
+  { appExpr  :: V.ID
     {- | I don't want to bother defining a BoolValue just for this;
         can be cleaned up in the future.
         Also, prob better to make the @args@ UBoolLit exprs or smtg
     -}
-    args :: [UBoolValue],
-    verDocId :: LSP.VersionedTextDocumentIdentifier
+  , args     :: [V.UBoolValue]
+  , verDocId :: LSP.VersionedTextDocumentIdentifier
   }
   deriving stock (Eq, Show, Generic)
   deriving anyclass (FromJSON, ToJSON)
 
 -- | l4/evalApp response payload
 newtype EvalAppResult = EvalAppResult
-  { value :: UBoolValue }
+  { value :: V.UBoolValue }
+  deriving stock (Eq, Show)
+
+instance ToJSON EvalAppResult where
+  toJSON (EvalAppResult val) = object [
+    "$type" .= ("EvalAppResult" :: T.Text),
+    "value" .= val
+    ]
+
+instance FromJSON EvalAppResult where
+  parseJSON = withObject "EvalAppResult" $ \obj -> do
+    typeTag <- obj .: "$type"
+    if typeTag == ("EvalAppResult" :: T.Text)
+      then EvalAppResult <$> obj .: "value"
+      else fail $ "Expected $type field to be 'EvalAppResult' but got: " <> T.unpack typeTag
```

Comment:
I'm guessing this happens because you cannot use the standard encoding on the other side?

Reply:
where `the = haskell's`?

Reply:
ah it's just because, strictly speaking, the interface / schema that the frontend expects has a $type field:

```
export const EvalAppResult = Schema.Struct({
  $type: Schema.tag('EvalAppResult'),
  value: BoolValue,
})
```

If there's a cleaner or more idiomatic way to make sure that the encoded json has that $type field, I'm happy to switch it to that way.

It would also be reasonable to say that we shouldn't require a $type field for this in the first place